pv211_utils/evaluation_metrics.py

```diff
@@ -0,0 +1,405 @@
+        current_rank += 1
+
+    precision = num_relevant / k
+
+    return precision
+
+
+def _calc_average_precision(system: IRSystemBase, judgements: Set, k: int,
+                            query: QueryBase) -> float:
+    num_relevant = 0
+    average_precision = 0.0
+    current_rank = 1
+
+    for document in system.search(query):
+        if current_rank > k:
+            break
+        if (query.query_id, document.document_id) in judgements:
+            num_relevant += 1
+            average_precision += num_relevant / current_rank
+        current_rank += 1
+
+    average_precision /= num_relevant if num_relevant > 0 else 1
+
+    return average_precision
+
+
+class calc_map(abc.ABC):
+    system = None
+    judgements = None
+    k = None
+    _CURRENT_INSTANCE = None
+
+    @classmethod
+    def _calc_average_precision(csl, query: QueryBase) -> float:
+        num_relevant = 0
+        average_precision = 0.0
+        current_rank = 1
+
+        for document in csl._CURRENT_INSTANCE.system.search(query):
+            if current_rank > csl._CURRENT_INSTANCE.k:
+                break
+            if (query.query_id, document.document_id) in csl._CURRENT_INSTANCE.judgements:
+                num_relevant += 1
+                average_precision += num_relevant / current_rank
+            current_rank += 1
+
+        average_precision /= num_relevant if num_relevant > 0 else 1
+
+        return average_precision
+
+    def mean_average_precision(self, system: IRSystemBase, queries: OrderedDict,
+                               judgements: Set[JudgementBase],
+                               k: int, num_processes: int) -> float:
+        """Evaluate system for given queries and judgements with mean average precision
+        metric. Where first k documents will be used in evaluation.
+
+        Arguments
+        ---------
+        system : IRSystemBase
+            System to be evaluated.
+        queries : OrderedDict
+            Queries to be searched.
+        judgements : Set[JudgementBase]
+            Judgements.
+        k : int
+            Parameter defining evaluation depth.
+        num_processes : int
+            Parallelization parameter defining number of processes to be used to run the evaluation.
+
+        Returns
+        -------
+        float
+            Mean average precision score from interval [0, 1].
+        """
+        map_score = 0.0
+
+        self.system = system
+        self.judgements = _judgements_obj_to_id(judgements)
+        self.k = k
+        self.__class__._CURRENT_INSTANCE = self
+
+        if num_processes == 1:
+            for query in list(queries.values()):
+                map_score += self.__class__._calc_average_precision(query)
+        else:
+            with get_context("fork").Pool(processes=num_processes) as process_pool:
+                for precision in process_pool.imap(self.__class__._calc_average_precision, list(queries.values())):
+                    map_score += precision
+
```

Comment:
The evaluation can take a bit. Let's be friendly to users and give them an ETA? We can do `from tqdm import tqdm` at the top and then do the following here:
```suggestion
        self.__class__._CURRENT_INSTANCE = self
        query_values = list(queries.values())
        query_values = tqdm(query_values)

        if num_processes == 1:
            for query in query_values:
                map_score += self.__class__._calc_average_precision(query)
        else:
            with get_context("fork").Pool(processes=num_processes) as process_pool:
                for precision in process_pool.imap(self.__class__._calc_average_precision, query_values):
                    map_score += precision

```

Here and in all top-level functions from the `evaluation_measures` module.